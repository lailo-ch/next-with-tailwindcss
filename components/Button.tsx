import * as React from "react";

const Button: React.FunctionComponent = ({ children }) => {
  return <button>{children}</button>;
};

export default Button;
