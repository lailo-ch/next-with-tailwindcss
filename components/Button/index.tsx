import * as React from 'react'
import cn from 'classnames'

interface Props {
  type?: 'primary' | 'secondary'
  className?: string | object
  loading?: boolean
  [key: string]: any
}

const Button: React.FunctionComponent<Props> = ({
  children,
  className,
  type = 'primary',
  loading = false,
  ...props
}) => {
  const disabled = props.disabled || loading

  const mergedClassName = cn(
    'px-4 py-2',
    { 'text-white': !disabled, 'text-gray-500': disabled },
    {
      'bg-green-700 hover:bg-green-800': !disabled && type === 'primary',
      'bg-blue-700 hover:bg-blue-800': !disabled && type === 'secondary',
      'bg-gray-300': disabled,
    },
    {
      'cursor-not-allowed': disabled && !loading,
      'cursor-wait': loading,
    },
    className
  )

  return (
    <button className={mergedClassName} disabled={disabled} {...props}>
      {children}
      {loading && '...'}
    </button>
  )
}

export default Button
