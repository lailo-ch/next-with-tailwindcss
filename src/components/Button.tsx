import * as React from 'react'
import clsx from 'clsx'

export interface Props {
  type?: 'primary' | 'secondary'
  className?: string | Record<string, unknown>
  loading?: boolean
}

const Button: React.FC<Props & React.HTMLProps<HTMLButtonElement>> = ({
  children,
  className,
  type = 'primary',
  loading = false,
  ...props
}) => {
  const disabled = props.disabled || loading

  const mergedClassName = clsx(
    'px-4 py-2',
    { 'text-white': !disabled, 'text-gray-500 dark:text-gray-700': disabled },
    {
      'bg-primary hover:bg-primary-darker': !disabled && type === 'primary',
      'bg-secondary hover:bg-secondary-darker':
        !disabled && type === 'secondary',
      'bg-gray-300 dark:bg-gray-800': disabled,
    },
    {
      'cursor-not-allowed': disabled && !loading,
      'cursor-wait': loading,
    },
    className
  )

  return (
    <button className={mergedClassName} disabled={disabled} {...props}>
      {children}
      {loading && '...'}
    </button>
  )
}

export default Button
