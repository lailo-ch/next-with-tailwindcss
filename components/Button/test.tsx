import * as React from 'react'
import { shallow } from 'enzyme'
import toJson from 'enzyme-to-json'

import Button from '.'

describe('Button', () => {
  it('renders with correct label', () => {
    const wrapper = shallow(<Button>Hello</Button>)
    expect(wrapper.text()).toEqual('Hello')
  })

  it('matches primary (default) type snapshot', () => {
    const wrapper = shallow(<Button>Primary</Button>)
    expect(toJson(wrapper)).toMatchSnapshot()
  })

  it('matches secondary type snapshot', () => {
    const wrapper = shallow(<Button type="secondary">Secondary</Button>)
    expect(toJson(wrapper)).toMatchSnapshot()
  })

  it('matches with external className as string', () => {
    const wrapper = shallow(
      <Button className="external-classname">Hello</Button>
    )
    expect(toJson(wrapper)).toMatchSnapshot()
  })

  it('matches with external className as object', () => {
    const wrapper = shallow(
      <Button
        className={{
          'external-classname-true': true,
          'external-classname-false': false,
        }}
      >
        Hello
      </Button>
    )
    expect(toJson(wrapper)).toMatchSnapshot()
  })

  it('matches disabled state', () => {
    const wrapper = shallow(<Button disabled>Disabled</Button>)
    expect(toJson(wrapper)).toMatchSnapshot()
  })

  it('matches loading state', () => {
    const wrapper = shallow(<Button loading>Disabled</Button>)
    expect(toJson(wrapper)).toMatchSnapshot()
  })
})
