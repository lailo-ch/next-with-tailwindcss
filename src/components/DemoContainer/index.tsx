import * as React from 'react'
import { Props as ButtonProps } from '../Button'

interface Props {
  Component: React.FunctionComponent
  title: string
}

const componentPopsList: Partial<ButtonProps>[] = [
  {
    type: 'primary',
    className: 'mr-4',
  },
  {
    type: 'secondary',
    className: 'mr-4',
  },
  {
    type: 'primary',
    className: 'mr-4',
    loading: true,
  },
  {
    type: 'secondary',
    disabled: true,
  },
]

const DemoContainer: React.FunctionComponent<Props> = ({
  Component,
  title,
}) => {
  return (
    <div className="bg-white my-5 p-5 shadow hover:shadow-xl">
      <h2 className="m-w-50 mb-4">{title}</h2>
      <div>
        {componentPopsList.map((props, index) => (
          <Component key={index} {...props}>
            {props.type}
          </Component>
        ))}
      </div>
    </div>
  )
}

export default DemoContainer
