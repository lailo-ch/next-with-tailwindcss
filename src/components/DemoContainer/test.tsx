import * as React from 'react'
import { shallow } from 'enzyme'
import toJson from 'enzyme-to-json'

import DemoContainer from '.'
import Button from '../Button'

describe('DemoContainer', () => {
  it('renders with button component', () => {
    const wrapper = shallow(<DemoContainer Component={Button} title="Button" />)
    expect(wrapper.find('h2').text()).toEqual('Button')
    expect(toJson(wrapper)).toMatchSnapshot()
  })
})
