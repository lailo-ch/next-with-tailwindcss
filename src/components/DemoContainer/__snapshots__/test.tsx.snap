// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DemoContainer renders with button component 1`] = `
<div
  className="bg-white my-5 p-5 shadow hover:shadow-xl"
>
  <h2
    className="m-w-50 mb-4 text-black"
  >
    Button
  </h2>
  <div>
    <Button
      className="mr-4"
      key="0"
      type="primary"
    >
      primary
    </Button>
    <Button
      className="mr-4"
      key="1"
      type="secondary"
    >
      secondary
    </Button>
    <Button
      className="mr-4"
      key="2"
      loading={true}
      type="primary"
    >
      primary
    </Button>
    <Button
      disabled={true}
      key="3"
      type="secondary"
    >
      secondary
    </Button>
  </div>
</div>
`;
